### DISCUSSION TITLE
SPARQLProtocolSession custom HTTP Header

### DISCUSSION THREAD

**Discussion by mariussl** [eclipse-rdf4j/rdf4j#4136] [Q&A] [answered]
I am using [SPARQLRepository ](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/repository/sparql/SPARQLRepository.html) to connect via HTTP to Stardog. [Stardog is configured to authenticate incoming connections via OAuth2.](https://docs.stardog.com/operating-stardog/security/oauth-integration#oauth-20-identity-providers)

So I need to set the required Authorization Header for every Request made to Stardog. I know of [SPARQLRepository.setAdditionalHttpHeaders](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/repository/sparql/SPARQLRepository.html#setAdditionalHttpHeaders-java.util.Map-) but this methods intention is to set custom http headers for all requests issued.

What I need is to set a custom http header for every [SPARQLProtocolSession](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/http/client/SPARQLProtocolSession.html) that is used by a [SPARQLTupleQuery](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/repository/sparql/query/SPARQLTupleQuery.html). Currently this is only possible by creating a new HttpClient for every SPARQLConnection and setting it through [SPARQLConnection.setHttpClient](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.html#setHttpClient-org.apache.http.client.HttpClient-). Codewise this looks like:

`HttpClientBuilder builder = new HttpClientBuilder()`
initialize builder
`BasicHeader header = BasicHeader(HttpHeaders.Authorization, "bearer JWTTOKEN")`
`builder.setDefaultHeaders(listOf(header))`
`sparqlConnection.httpClient = builder.build()`

Then the [SPARQLTupleQuery](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/repository/sparql/query/SPARQLTupleQuery.html) spawned by [SPARQLConnection.prepareTupleQuery](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.html#prepareTupleQuery-org.eclipse.rdf4j.query.QueryLanguage-java.lang.String-java.lang.String-) will receive my custom HttpClient and pass it on to [SPARQLTupleQuery.evaluate](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/repository/sparql/query/SPARQLTupleQuery.html#evaluate--) as SPARQLProtocolSession. 

The downside of this is that I need to create a new HttpClientBuilder and HttpClient for every SPARQLConnection, so for every request issued to Stardog. I would prefer it if I could set the custom http header directly on the [SPARQLProtocolSession](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/http/client/SPARQLProtocolSession.html) used by [SPARQLTupleQuery](https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/repository/sparql/query/SPARQLTupleQuery.html) . 

I have not yet found a public API to do this. Does anybody know of a way how to do this or has some other advice?

**Comment by ate47** [comment, marked as answer]
You can use an interceptor to add your auth header in your HttpClient, I think this code can help you:

```java
HttpClientBuilder bld = HttpClientBuilder.create();
bld.addInterceptorFirst(
    (HttpRequestInterceptor) (httpRequest, httpContext) ->
        httpRequest.addHeader(HttpHeaders.AUTHORIZATION, "bearer JWTTOKEN")
);
CloseableHttpClient httpClient = bld.build();
```

You can then replace your auth header without having to recreate the HttpClient
